File: ink_engine/ink_engine/core/tiers.py
"""模型分层挡位原语（挡位配置模型 / 按挡位建链 / 调用统计钩子）。

模型分层：router/main 双挡位（轻量/主）按场景分配——
路由决策用轻量挡，内容生成用主挡。**组 → 挡位映射**属宿主
业务语义（TextForge 的 ``GROUP_MODEL_TIER``），引擎只提供机制：

- :func:`tier_key`：挡位名 → 配置键前缀（未知/缺省回落 main）；
- :func:`resolve_tier_config`：从用户模型配置解析单挡位的「主配置 + 备用列表」；
- :func:`build_tier_chain`：按挡位构建 ``ModelChain``（复用重试/备用链）；
- :class:`TierCallStats`：挡位调用统计钩子（``llm_calls_by_tier`` 观测）。

配置形态：``model_config`` 为 dict，挡位配置键 =
``f"{tier}_config"``（缺省回退 ``main_config``），备用链键 =
``f"{tier}_fallback_configs"``（兼容历史嵌套 ``fallback_configs``）。

依赖注入：``build_tier_chain`` 的 ``create`` / ``retry`` 可注入（测试用假
模型 / 自定义重试），默认走引擎 ``create_llm`` + 标准重试。
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

# 默认挡位声明（装配注入前的出厂形态：main/router 双挡；tiers.json
# 等外部声明经 set_tier_names 装配注入——策略选择随数据声明演化）
TIER_NAMES: tuple[str, ...] = ("main", "router")

# 生效中的挡位声明（数据驱动：tier_key/resolve_tier_config/统计
# 全部读此值；默认 = 出厂双挡，未知挡位回落 main 行为不变）
_ACTIVE_TIER_NAMES: tuple[str, ...] = TIER_NAMES

# 未知挡位的回落：任何未知/None 挡位按主挡位处理（配置兜底语义）
_DEFAULT_TIER = "main"


def current_tier_names() -> tuple[str, ...]:
    """当前生效的挡位声明（观察侧；装配注入后立即反映）。"""
    return _ACTIVE_TIER_NAMES


def set_tier_names(names: Any) -> None:
    """装配注入：以数据声明挡位集合（tiers.json 等外部配置读取后调用）。

    声明即权威（整组替换，不做增量追加——配置方看到的就是生效的）；
    ``main`` 必须存在（未知挡位回落锚点，缺失会被坏配置静默换挡）；
    校验直过，非法声明显式拒绝。

    Args:
        names: 挡位名序列（如 ("main", "router", "audit")）。

    Raises:
        ValueError: 空/重复/空字符串/缺 main。
    """
    global _ACTIVE_TIER_NAMES
    normalized = tuple(str(name) for name in names)
    if not normalized:
        raise ValueError("挡位声明不能为空（至少须含 main）")
    if len(set(normalized)) != len(normalized):
        raise ValueError(f"挡位声明含重复项: {normalized}")
    if any(not name for name in normalized):
        raise ValueError("挡位名不能为空字符串")
    if _DEFAULT_TIER not in normalized:
        raise ValueError(f"挡位声明缺回落锚点 {_DEFAULT_TIER!r}: {normalized}")
    _ACTIVE_TIER_NAMES = normalized


def tier_key(tier: str | None) -> str:
    """挡位名 → 配置键前缀；未知或 None 回落 main（防拼写错误静默换挡）。"""
    return tier if tier in _ACTIVE_TIER_NAMES else _DEFAULT_TIER


@dataclass(frozen=True, slots=True)
class TierConfig:
    """单挡位的模型配置形态。

    Attributes:
        tier: 归一化后的挡位名（未知已回落 main）。
        config: 主配置 dict；None = 该挡位无配置（调用方走错误兜底）。
        fallbacks: 备用配置列表（与主配置同形态，ModelChain 备用链）。
    """

    tier: str
    config: dict[str, Any] | None
    fallbacks: tuple[dict[str, Any], ...]


def resolve_tier_config(
    model_config: dict[str, Any] | None,
    tier: str | None,
) -> TierConfig:
    """从用户模型配置解析指定挡位的配置形态（纯函数，可单测）。

    解析规则（与宿主 `_build_tier_chain` 原语义一致，单点维护）：
    1. 主配置 = ``f"{tier}_config"``，缺省回退 ``main_config``；
    2. 备用列表 = ``f"{tier}_fallback_configs"``，兼容历史嵌套
       ``config["fallback_configs"]``；
    3. 全部缺失 → ``config=None``（调用方按无配置处理，不抛错）。
    """
    key = tier_key(tier)
    cfg_map = model_config or {}
    # 显式空配置与缺失区分（ENG3-11）：``or`` 回退会把显式空配置
    # ``{f"{key}_config": {}}`` 误当缺失而回落主挡位——空配置 = 该挡位
    # 显式声明无配置（config=None，调用方按无配置兜底），不回落
    cfg = cfg_map.get(f"{key}_config")
    if cfg is None:
        cfg = cfg_map.get("main_config")
    if cfg is None:
        cfg = {}
    tier_fallbacks = cfg_map.get(f"{key}_fallback_configs")
    if tier_fallbacks is None:
        tier_fallbacks = cfg.get("fallback_configs")
    fallbacks = tier_fallbacks if tier_fallbacks is not None else []
    return TierConfig(
        tier=key,
        config=cfg or None,
        fallbacks=tuple(fallbacks or ()),
    )


def build_tier_chain(
    model_config: dict[str, Any] | None,
    tier: str | None = None,
    *,
    create: Any = None,
    retry: Any = None,
):
    """按挡位构建模型链（主配置 + 备用链）；配置缺失返回 None。

    Args:
        model_config: 用户模型配置字典（可含挡位主配置与备用列表）。
        tier: 挡位名（router/tool/main/audit；未知回落 main）。
        create: 适配器工厂注入（默认引擎 create_llm；测试注入假模型）。
        retry: 重试策略注入（默认引擎标准重试）。

    Returns:
        ``ModelChain`` 实例；该挡位与主挡位均无配置时返回 None
        （调用方按配置缺失兜底，与引擎节点容错语义一致）。
    """
    from ink_engine.core.llm.fallback import ModelChain

    resolved = resolve_tier_config(model_config, tier)
    if not resolved.config:
        return None
    return ModelChain(
        [resolved.config, *resolved.fallbacks],
        create=create,
        retry=retry,
    )


class TierCallStats:
    """挡位调用统计钩子：按挡位累加 LLM 调用次数，供回合级观测。

    用法（宿主回合收尾）：每次按挡位发起调用后 ``record(tier)``，
    回合结束 ``snapshot()`` 汇入 ``turn_metrics.llm_calls_by_tier``。
    线程安全：单回合单执行流内使用（与引擎执行语义一致，无锁）。
    """

    def __init__(self) -> None:
        self._counts: dict[str, int] = {}

    def record(self, tier: str | None, count: int = 1) -> None:
        """累加一次（或多次）某挡位的调用数；未知挡位归一后记录。"""
        if count <= 0:
            return
        key = tier_key(tier)
        self._counts[key] = self._counts.get(key, 0) + count

    def snapshot(self) -> dict[str, int]:
        """当前计数快照（{挡位: 次数}，未调用过的挡位不出现）。"""
        return dict(self._counts)

    def reset(self) -> None:
        """清零（新回合复用实例时调用）。"""
        self._counts.clear()

    def __iadd__(self, other: TierCallStats) -> TierCallStats:
        """合并另一实例的计数（嵌套图/子图回流场景汇总）。"""
        for tier, count in other._counts.items():
            self._counts[tier] = self._counts.get(tier, 0) + count
        return self


__all__ = [
    "TIER_NAMES",
    "TierCallStats",
    "TierConfig",
    "build_tier_chain",
    "current_tier_names",
    "resolve_tier_config",
    "set_tier_names",
    "tier_key",
]

File: ink-ts/engine/src/core/tiers/tiers.ts
/**
 * 模型分层挡位原语（挡位配置模型 / 按挡位建链 / 调用统计钩子）——tiers.py 移植。
 *
 * 模型分层：router/main 双挡位（轻量/主）按场景分配——路由决策用轻量挡，
 * 内容生成用主挡。组 → 挡位映射属宿主业务语义，引擎只提供机制：
 * 挡位名 → 配置键前缀、单挡位配置解析、按挡位建链与挡位调用统计。
 *
 * 配置形态：model_config 为 dict，挡位配置键 = f"{tier}_config"（缺省回退
 * main_config），备用链键 = f"{tier}_fallback_configs"（兼容历史嵌套
 * fallback_configs）。显式空配置（{}）与缺失键区分：前者 = 该挡位显式
 * 声明无配置（不回落主挡位），后者 = 回落 main。
 *
 * 依赖注入：build_tier_chain 的 create / retry 可注入；ModelChain 的
 * 重试/备用/流式执行属 llm 层（未移植），本模块只负责按挡位装配有序的
 * 模型配置清单（configs = [主配置, ...备用]），llm 层移植后接线。
 */

import { isRecord, type JsonRecord } from '../json.js';

/** 默认挡位声明（装配注入前的出厂形态：main/router 双挡）。 */
export const TIER_NAMES = ['main', 'router'] as const;

/** 未知挡位的回落锚点：任何未知/None 挡位按主挡位处理（配置兜底语义）。 */
const DEFAULT_TIER = 'main';

/** 生效中的挡位声明（数据驱动：set_tier_names 整组替换，校验直过）。 */
let activeTierNames: readonly string[] = TIER_NAMES;

/** Python str() 口径的标量渲染（挡位声明归一化；None → 'None'）。 */
function nameStr(value: unknown): string {
  if (value === null || value === undefined) return 'None';
  if (typeof value === 'boolean') return value ? 'True' : 'False';
  if (typeof value === 'string') return value;
  return String(value);
}

/** Python tuple repr 口径（错误消息携带声明清单，便于定位）。 */
function tupleRepr(names: readonly string[]): string {
  return `(${names.map((name) => `'${name}'`).join(', ')})`;
}

/** 当前生效的挡位声明（观察侧；装配注入后立即反映；拷贝防外部污染）。 */
export function current_tier_names(): readonly string[] {
  return [...activeTierNames];
}

/** 装配注入：以数据声明挡位集合；声明即权威（整组替换），main 必须存在。 */
export function set_tier_names(names: readonly unknown[]): void {
  const normalized = names.map(nameStr);
  if (normalized.length === 0) {
    throw new Error('挡位声明不能为空（至少须含 main）');
  }
  if (new Set(normalized).size !== normalized.length) {
    throw new Error(`挡位声明含重复项: ${tupleRepr(normalized)}`);
  }
  if (normalized.some((name) => name === '')) {
    throw new Error('挡位名不能为空字符串');
  }
  if (!normalized.includes(DEFAULT_TIER)) {
    throw new Error(`挡位声明缺回落锚点 '${DEFAULT_TIER}': ${tupleRepr(normalized)}`);
  }
  activeTierNames = normalized;
}

/** 挡位名 → 配置键前缀；未知或 None 回落 main（防拼写错误静默换挡）。 */
export function tier_key(tier: string | null | undefined): string {
  return typeof tier === 'string' && activeTierNames.includes(tier) ? tier : DEFAULT_TIER;
}

/** 单挡位的模型配置形态（frozen 数据：tier/config/fallbacks）。 */
export class TierConfig {
  readonly tier: string;
  readonly config: JsonRecord | null;
  readonly fallbacks: readonly JsonRecord[];

  constructor(tier: string, config: JsonRecord | null, fallbacks: readonly JsonRecord[]) {
    this.tier = tier;
    this.config = config;
    this.fallbacks = fallbacks;
  }
}

function recordOf(value: unknown): JsonRecord | null {
  return isRecord(value) ? (value as JsonRecord) : null;
}

/**
 * 从用户模型配置解析指定挡位的配置形态（纯函数）。
 * 主配置 = f"{tier}_config"，缺省回退 main_config；备用列表 =
 * f"{tier}_fallback_configs"，兼容历史嵌套 config["fallback_configs"]；
 * 全部缺失 → config=null（调用方按无配置处理，不抛错）。
 */
export function resolve_tier_config(
  model_config: JsonRecord | null | undefined,
  tier: string | null | undefined,
): TierConfig {
  const key = tier_key(tier);
  const cfgMap = model_config ?? {};
  let cfg: unknown = cfgMap[`${key}_config`];
  if (cfg === null || cfg === undefined) cfg = cfgMap['main_config'];
  if (cfg === null || cfg === undefined) cfg = {};
  const cfgRecord = recordOf(cfg);
  let tierFallbacks: unknown = cfgMap[`${key}_fallback_configs`];
  if (tierFallbacks === null || tierFallbacks === undefined) {
    tierFallbacks = cfgRecord === null ? undefined : cfgRecord['fallback_configs'];
  }
  const fallbackList = Array.isArray(tierFallbacks) ? tierFallbacks : [];
  const config = cfgRecord !== null && Object.keys(cfgRecord).length > 0 ? cfgRecord : null;
  return new TierConfig(key, config, [...(fallbackList as JsonRecord[])]);
}

/** 链装配结果：ModelChain 的 configs 观察面（主配置在前，备用随后；类型不出模块）。 */
interface ModelChain {
  readonly configs: readonly JsonRecord[];
}

/**
 * 按挡位构建模型链（主配置 + 备用链）；配置缺失返回 null。
 * 该挡位与主挡位均无配置时返回 null（调用方按配置缺失兜底）。
 * create/retry 为 llm 层 seam（llm 未移植，仅保留注入面，暂不消费）。
 */
export function build_tier_chain(
  model_config: JsonRecord | null | undefined,
  tier: string | null = null,
  options: { create?: (config: JsonRecord) => unknown; retry?: unknown } = {},
): ModelChain | null {
  void options;
  const resolved = resolve_tier_config(model_config, tier);
  if (resolved.config === null) return null;
  return { configs: [resolved.config, ...resolved.fallbacks] };
}

/** 挡位调用统计钩子：按挡位累加 LLM 调用次数，供回合级观测。 */
export class TierCallStats {
  private counts = new Map<string, number>();

  /** 累加一次（或多次）某挡位的调用数；未知挡位归一后记录，非正计数忽略。 */
  record(tier: string | null | undefined, count = 1): void {
    if (count <= 0) return;
    const key = tier_key(tier);
    this.counts.set(key, (this.counts.get(key) ?? 0) + count);
  }

  /** 当前计数快照（{挡位: 次数}，未调用过的挡位不出现）。 */
  snapshot(): Record<string, number> {
    const out: Record<string, number> = {};
    for (const [tier, count] of this.counts) out[tier] = count;
    return out;
  }

  /** 清零（新回合复用实例时调用）。 */
  reset(): void {
    this.counts.clear();
  }

  /** 合并另一实例的计数（+= 语义；嵌套图/子图回流场景汇总），返回自身。 */
  merge(other: TierCallStats): this {
    for (const [tier, count] of other.counts) {
      this.counts.set(tier, (this.counts.get(tier) ?? 0) + count);
    }
    return this;
  }
}
